### DISCUSSION TITLE
eglot-workspace-configuration can *only* be global or dir-local?

### DISCUSSION THREAD

**Discussion by phil-s** [joaotavora/eglot#1077] [General]
The manual says:

> Note that while it is possible to set this variable globally **or buffer-locally**, doing so makes little sense. It is usually set via .dir-locals.el or [special-purpose Elisp functions](https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Variables.html).

I think that text needs a big caveat attached to it?

I see that this was discussed at length in https://github.com/joaotavora/eglot/discussions/967 but to review in brief:

```el
(cl-defmethod eglot-handle-request
  (server (_method (eql workspace/configuration)) &key items)
  "Handle server request workspace/configuration."
...
            (with-temp-buffer
...
                (setq-local major-mode (car (eglot--major-modes server)))
                (hack-dir-local-variables-non-file-buffer)
                (cl-loop for (wsection o)
                         on (eglot--workspace-configuration-plist server)
...
```

With that latter function call just returning `eglot-workspace-configuration` unless the user has defined a function named `eglot-workspace-configuration` by the looks of it?

Notwithstanding the custom function option, setting a buffer-local value for `eglot-workspace-configuration` in the major mode hook for the language (i.e. the typical way of setting a buffer-local value) can't work with this arrangement.

I feel that in certain cases it *should* be absolutely fine to want to do this in a mode hook; but as that's not supported at present I think the manual needs to make it *much* clearer that the directory-local approach is the *only* "buffer-local" approach which will work.

**Comment by joaotavora** [comment]
> I feel that in certain cases it should be absolutely fine to want to do this in a mode hook; but as that's not supported at present I think the manual needs to make it much clearer that the directory-local approach is the only "buffer-local" approach which will work.

You "feel" that, but as the manual says, it "makes little sense".  Are you hanging on to that residue of sense that you think it makes?

Anyway, there's a section devoted to this already, I don't think we need any more? Besides, Eli Zaretskii is working on a rewrite of the manual to be placed in the core, so I won't be making any changes to `MANUAL.md` while that is ongoing.  Once the manual is in Emacs.git anyone can propose changes to it.

**Comment by phil-s** [comment]
All good; I can follow up with Eli.

> You "feel" that, but as the manual says, it "makes little sense". Are you hanging on to that residue of sense that you think it makes?

My current understanding (which I acknowledge may be wrong or otherwise lacking) is that the reasons why it "makes little sense" are:

1. When a `workspace/configuration` request occurs, there might not be any relevant buffers at all.
2. If buffers can have different buffer-local values, then the responses to the language server could be variable.

For (1) I would argue that as there is surely(?) a relevant buffer when `eglot` starts, it could obtain the configuration at that time from that buffer, and remember it.  And (2) is only an issue at all if the user is modifying the config, and even then they could trivially restart the server or otherwise trigger a config change in a new/reverted buffer after making a change, and once again that could simply be remembered for later.

The benefit being that setting buffer-local values in mode hooks is very common and well-understood, whereas the current mechanism with the temp buffer is very surprising and going to cause confusion (unless it's clearly pointed out, which is not currently the case).

I'm not convinced that I'm right, mind; I merely want things to be easier than they are at the moment, one way or another (said after having spent many, *many* hours over the past couple of days trying to convince a language server to behave the way I want -- or even just consistently -- and finding the documentation (for both the server and the client) to be much more confusing than I'd hoped for).

**Comment by joaotavora** [comment]
> The benefit being that setting buffer-local values in mode hooks is very common and well-understood, whereas the current mechanism with the temp buffer is very surprising and going to cause confusion (unless it's clearly pointed out, which is not currently the case).

But that mechanism is an implementation detail.  You shouldn't need to look at it, as long as you understand that this is a per-project setting by definition (LSP parlance these are "workspaces", hence the name).  In Emacs' `project.el` projects are represented by "root directories". Hence, directory-local is the correct abstraction to use here.  The fact that project-locality is implemented in terms of buffer-locality is, again, an implementation detail.  It could be implemented some other way, but this way was the easiest one.

> I merely want things to be easier than they are at the moment

What exactly was not easy?  The manual has IMO clear examples of how to set directory-local values for this variable, both in `.dir-locals.el` and using Elisp mechanisms.

Can you post the solution you ended up with?  Maybe I can help simplify it and add the example to the manual, presuming it adds something that is not already there.

**Reply by joaotavora** [reply]
> And (2) is only an issue at all if the user is modifying the config

Not so.  Even if the user doesn't modify it during the course of the connection, a server serving two different major modes could easily be erroneously be sent two different workspace configurations depending on whether the `M-x eglot` happened in a buffer of one major mode or the other.  This is unwanted.  Thus the safe and easy way to fix this is again to ensure the variable is set in a dir-local way.  For example, in `.dir-locals.el` of your project's root.

```
((nil
  . ((eglot-workspace-configuration
      . (:clojure-lsp (:someclojurelsp-specific-setting 42))))))
```

In this example, I used "clojure-lsp" because that server is known to support `clojure-mode`, `clojurescript-mode` and `clojurec-mode` major modes.

**Comment by phil-s** [comment]
> What exactly was not easy?

Bearing in mind that I started looking at this before the recent commits and documentation updates regarding `eglot-workspace-configuration`, so there have been many changes in this area in the interim.  The alist vs plist change was very confusing for me at one point (it's possible that I was using 1.8 and reading about 1.9).

Part of the problem is that the language servers are dealing with -- and documenting -- JSON, and eglot is documenting plists, and I saw no documented process for correctly converting what would inevitably be a JSON starting point into the *correct* plist (given the knobs for varying how certain JSON values are handled), which was needed in order to have any confidence that I would be sending the server what I intended to sent (the new command for showing some JSON is a significant improvement in this regard).  After digging into the eglot code I found `jsonrpc--json-read` and came up with the following as a method of producing that plist:

```el
(pp-display-expression (jsonrpc--json-read) "*json*")
{
    JSON HERE
}
```

I think there should either be a documented process for doing this (ideally without calling a 'private' function), or otherwise a way to simply pass a JSON string in the first place, because I think that's what most people will actually have to begin with.  And if the command for showing the config as JSON *isn't* showing what they expect, users need to know how to address that.

(Edit: Looking again at the updated manual I see that the specific syntax required did actually get documented in the recent changes.  I nevertheless believe that some kind of integrated converter would be helpful for users.)

(n.b. What that `JSON HERE` is *actually* supposed to be in the above pseudo-data is still a mystery to me.  See https://github.com/bmewburn/intelephense-docs/issues/50 regarding that.)

The buffer-local thing genuinely threw me, because it never occurred to me that eglot wasn't simply obtaining the values from one of the file-visiting buffers (again, the temp buffer thing is really unexpected), and so it never occurred to me that there would be a difference between setting that in .dir-locals.el vs in a mode hook.  I'd started with a .dir-locals.el file, then later moved this particular variable to a mode hook, and it took me a long time to realise that removing the .dir-locals.el version had caused any breakage.

I might well have noticed sooner, but I've been struggling with other failures as well, so this was subtle enough to get lost in the noise of things like the server dying almost immediately after connection (despite working fine earlier in the day).  Which means...

> Can you post the solution you ended up with?

...I haven't ended up with anything I'd call "a solution" yet.

I also struggled with `eglot-server-programs` on account of this:

```
(let ((eglot-server-programs
       '((((drupal-mode :language-id "php")
           php-mode)
          "intelephense" "--stdio"))))
  (eglot--lookup-mode 'drupal-mode))

=> ((drupal-mode php-mode) "php" ("intelephense" "--stdio"))


(let ((eglot-server-programs
       '(((php-mode
           (drupal-mode :language-id "php"))
          "intelephense" "--stdio"))))
  (eglot--lookup-mode 'drupal-mode))

=> ((drupal-mode php-mode) "drupal" ("intelephense" "--stdio"))
```

I was initially using the latter, and eventually spotted in the events log that the language server was complaining about the language ID being "drupal" despite the fact that I'd set it to "php".  This seems like a bug?

Anyhow, this is my first foray into using LSP, and when I don't know for sure what the server wants, and I'm also experiencing difficulties making the client send what I thought it was sending, and the visible data format is not JSON, I've been finding debugging to be slow going; so any and all improvements to docs and processes are a win from my perspective.

**Reply by phil-s** [reply]
> and the visible data format is not JSON

On that note, I was thinking that `jsonrpc` might benefit from an option to actually log JSON.  It presumably sees the JSON for incoming requests and responses, and it certainly has the JSON at the time it sends its own requests and responses, so why not have an optional log buffer in that format?

**Reply by joaotavora** [reply]
Sorry, but this started as a misunderstanding or difficulty in relation to the documentation of `eglot-workspace-configuration` and the issues of buffer-locality. In a cursory reading of your long reply, I don't see anything really relevant to it anymore.  Is the documentation about this specific topic present in MANUAL.md satisfactory to you right now?  The MANUAL.md always describes the latest version of Eglot. 

> again, the temp buffer thing is really unexpected

I've told you: it's an implementation detail.  It's not documented in the MANUAL.md, so you shouldn't really "expect" anything from Eglot in that regard. If you follow the MANUAL.md instead of trying to guess from the source, the program should do what the manual says it does.  The manual says you should set directory-local values of `eglot-workspace-configuration`, so I suggest you try that.

Now you seem to be addressing multiple different topics, including a possible bug with `eglot-server-programs` and `:language-id`, the serialization of Elisp objects into JSON objects, and the Eglot events buffer.  You're suggesting that use of a private function is needed to solve your problem, and I don't understand that either.  So this is becoming hard to follow.

I think you should make different postings for each of these problems in reproducible fashion, in accordance with https://github.com/joaotavora/eglot/blob/master/MANUAL.md#reporting-bugs.  If you prefer, you can make bug reports to Emacs's bug tracker using `M-x report-emacs-bug`, provided you follow those guidelines and Cc me in the report. 

If you think the MANUAL.md is currently wrong or inaccurate, please say where and how.

**Reply by phil-s** [reply]
I think the part I originally quoted is, to all practical purposes, incorrect:

> Note that while it is possible to set this variable globally or buffer-locally, doing so makes little sense.

As far as I can see, there's no practical way in which this variable can be set buffer-locally.  I would rephrase this as:

> Note that setting this variable buffer-locally will have no effect.  It is possible to set the global value instead of using a directory-local value, but in most cases that makes little sense.

I say "in most cases" as I have lots of single-purpose development environments which contain their own Emacs instance, and setting the value globally would be totally fine in one of those, as that instance of Emacs is never going to engage with any other programming environment.

And given that it can't be set buffer-locally, I'm a bit doubtful that it should be defined as defvar-local.  It will acquire buffer-local values via dir-locals regardless, but for something that is only useful as a dir-local or global value, I think defvar maybe makes better sense?

**Reply by phil-s** [reply]
I agree that the previous reply was a bit rambling.  You'd asked what it was that hadn't been easy, though (I've now added that as a quote at the top), so I was just answering that, but I can raise separate bug reports / feature requests for sure.

**Comment by joaotavora** [comment]
> Note that setting this variable buffer-locally will have no effect. It is possible to set the global value instead of using a directory-local value, but in most cases that makes little sense.

Something like that may be acceptable, with buffer-locally replaced by "in the major mode hook", but also think you should wait for Eli's version of the manual.  I get it that `defvar` would work, but I want to stress in that declaration that this variable will take on buffer-local values and that, as you well say, setting it globally generally makes little sense.
